### ISSUE TITLE
Typeability Bug in Reference Implementation

### ISSUE THREAD

**Issue by RossTate** [WebAssembly/reference-types#111]
If I'm reading [this](https://github.com/WebAssembly/reference-types/blob/master/interpreter/valid/valid.ml#L216-L221) code correctly, the reference interpretation incorrectly accepts this invalid instruction sequence:
```
(unreachable)
(select)
(ref.is_null funcref) ;; specific reference type doesn't matter
```

The issue is that `select` is polymorphic *with respect to numeric types only*. So while the result type of `select` after `unreachable` is unknown, it must at least be a numeric type. As such, it cannot be used for any instruction expecting a reference type.

The reference interpreter seems to fail to incorporate this reasoning. Instead of using `value_type option`, the reference interpreter seems to need to use an ADT with an additional nullary case that represents only unknown *numeric* types.

**Comment by rossberg** [comment]
Good catch. This actually is another fall-out from our removal of subtyping. :(  (Formerly, unreachable was typed with bottom, such that it was correct to accept this.)

**Comment by RossTate** [comment]
Oh whoa, I never even realized there had been a bottom *value* type. That causes problems as well.

**Comment by rossberg** [comment]
It was purely a spec-internal type, not expressible in the program itself. I actually mentioned it to you some time ago, and you didn't see any issues. See my presentation from the June '19 F2F meeting for the motivation.

**Comment by RossTate** [comment]
I understand. There's a difference between a bottom value type and a bottom state/result type. I thought you were referring to the latter concept. Sorry for the misunderstanding.

**Comment by rossberg** [comment]
Closed via #116.
